### ISSUE TITLE
Statistics Latvia - RUI020m - bad request

### ISSUE THREAD

**Issue by mapdek** [rOpenGov/pxweb#278]
Dear All,

I am trying to extract data from [this table from Statistics Latvia ](https://data.stat.gov.lv/pxweb/en/OSP_PUB/START__NOZ__RU__RUI/RUI020m/) using PXweb, but receive an error Bad request (HTPPP 400), please see the image below:

Is there something wrong on the side of the syntax?

```
library(pxweb)

pxweb_query_list <-
  list("NACE_MIG" = c("B_C_D"),
       "ContentsCode" = c("RUI020m"),
       "TIME" = c("*"))
 
px_data <-
  pxweb_get(url = "https://data.stat.gov.lv/api/v1/en/OSP_PUB/NOZ/RU/RUI/RUI020m",
            query = pxweb_query_list)  
```

![image](https://github.com/rOpenGov/pxweb/assets/106142419/43251223-e8dd-4818-a738-5a0cdfb2c2ab)

**Comment by pitkant** [comment]
@mapdek It seems that currently the query should look more like this:

```
pxweb_query_list <-
  list("NACE_MIG" = c("B_C_D_X_D353"),
       "ContentsCode" = c("RUI020m"),
       "TIME" = c("*"))
```

but I got the same error as you.

In the web page there is the following section:

"Api-V2:
Send a GET request to the URL below to access this table."

with the URL

https://api.stat.gov.lv/api/v2/tables/RUI020m/data?lang=en&valueCodes[NACE_MIG]=B_C_D_X_D353&valueCodes[ContentsCode]=RUI020m&valueCodes[TIME]=2025M08&codelist[NACE_MIG]=vs_VS_NACE2_BCD

which downloads the correct .px dataset.

The JSON query from the webpage looks like this:

```
{
  "query": [
    {
      "code": "NACE_MIG",
      "selection": {
        "filter": "vs:VS_NACE2_BCD",
        "values": [
          "B_C_D_X_D353"
        ]
      }
    },
    {
      "code": "ContentsCode",
      "selection": {
        "filter": "item",
        "values": [
          "RUI020m"
        ]
      }
    },
    {
      "code": "TIME",
      "selection": {
        "filter": "item",
        "values": [
          "2025M08"
        ]
      }
    }
  ],
  "response": {
    "format": "json-stat2"
  }
}
```

and immediately I am wondering about why "code": "NACE_MIG" has a "filter": "vs:VS_NACE2_BCD" instead of the regularly seen "filter": "item". Maybe this is the cause? When I use the json_query as input to pxweb_query and download the data

```
json_query_path <- "path/to/query.json"
pxq2 <- pxweb_query(json_query_path)
px_data <- pxweb_get(url = "https://data.stat.gov.lv/api/v1/en/OSP_PUB/START/NOZ/RU/RUI/RUI020m", query = pxq2)
```

However, for some reason `pxweb_advanced_get()` and `pxweb_c()` don't assign the class `pxweb_data` to this `json` class object making the call to method `as.data.frame.pxweb_data` fail. @MansMeg any idea why this would be in this case, compared to other json examples?

**Comment by MansMeg** [comment]
@mapdek - did this solve your issue?

**Comment by mapdek** [comment]
Dear @MansMeg , I tried again using the same code as above, but still the same http 400 error.
As a turn-around option I used the .R library selenider to simulate a webbrowser to obtain the data. I would still be interested to know why it doesn't work with pxweb

```
library(pxweb)

pxweb_query_list <-
  list("NACE_MIG" = c("B_C_D"),
       "ContentsCode" = c("RUI020m"),
       "TIME" = c("*"))
 
px_data <-
  pxweb_get(url = "https://data.stat.gov.lv/api/v1/en/OSP_PUB/NOZ/RU/RUI/RUI020m",
            query = pxweb_query_list)  
```

**Comment by MansMeg** [comment]
Hi!

I get the following:
```
> library(pxweb)
> pxweb_query_list <-
+   list("NACE_MIG" = c("B_C_D"),
+        "ContentsCode" = c("RUI020m"),
+        "TIME" = c("*"))
> px_data <-
+   pxweb_get(url = "https://data.stat.gov.lv/api/v1/en/OSP_PUB/NOZ/RU/RUI/RUI020m",
+             query = pxweb_query_list)  
Error in pxweb_validate_query_with_metadata(pxq, pxmd) : 
  Assertion on 'NACE_MIG' failed: Must be a subset of {'MIG_ING','MIG_NRG_XE','MIG_CAG','MIG_DCOG','MIG_NDCOG','B_C_D_X_D353','B','C','C10','C101','C102','C103','C105','C106','C107','C108','C11','C13','C14','C15','C16','C161','C162','C17','C18','C20','C21','C22','C23','C24','C25','C26','C27','C28','C29','C30','C31','C32','C33','D_X_D353'}, but has additional elements {'B_C_D'}.

> packageVersion("pxweb")
[1] ‘0.17.0’
```

Hence, it seems to be something with the query list?

**Comment by mapdek** [comment]
Hi @MansMeg , thanks for this! Using the interactive menu from PX library to create the query, these are the steps I take:

```
### load package
library(pxweb)

### start the interactive menu from PX library to create query
#d <- pxweb_interactive() 
```
 
<img width="916" height="698" alt="Image" src="https://github.com/user-attachments/assets/1dff4966-757c-455e-b431-12762623bf65" />

<img width="413" height="156" alt="Image" src="https://github.com/user-attachments/assets/3158ca6a-2500-449b-9c97-156c9fc03f1e" />

<img width="450" height="203" alt="Image" src="https://github.com/user-attachments/assets/c7a5b460-ffa5-4de7-924e-2c7dc24efbb6" />

<img width="481" height="214" alt="Image" src="https://github.com/user-attachments/assets/730c1261-bbdd-4fde-a4f7-f9db2e577a77" />

<img width="511" height="377" alt="Image" src="https://github.com/user-attachments/assets/d0ce4e7f-b02c-4692-a5e0-97476ac1d387" />

<img width="522" height="381" alt="Image" src="https://github.com/user-attachments/assets/d9c5045c-5ec3-422e-8ee0-4af1135b0a8c" />

<img width="794" height="220" alt="Image" src="https://github.com/user-attachments/assets/1a2a2c9f-f4d3-453e-8202-f5edd476b715" />

<img width="940" height="156" alt="Image" src="https://github.com/user-attachments/assets/23033862-62e9-4834-b4c6-9c0facd77054" />

<img width="940" height="371" alt="Image" src="https://github.com/user-attachments/assets/64bf4821-1121-4e5f-8bde-92a25f6d981d" />

<img width="940" height="155" alt="Image" src="https://github.com/user-attachments/assets/6bfa4903-1716-4893-b7b3-a53825062c14" />

<img width="940" height="193" alt="Image" src="https://github.com/user-attachments/assets/a0ba8f45-474e-4d55-ae08-ae6e88639e5a" />

And these are the final steps to obtain the data:

<img width="940" height="389" alt="Image" src="https://github.com/user-attachments/assets/5aff8482-2248-45c8-9772-02325ee5f33e" />

Maybe one thing to note it that on the website of this table on the Statistics Latvia website [https://data.stat.gov.lv/pxweb/en/OSP_PUB/START__NOZ__RU__RUI/RUI020m/](https://data.stat.gov.lv/pxweb/en/OSP_PUB/START__NOZ__RU__RUI/RUI020m/), there is this option below **Sector (NACE Rev. 2)/ Main industrial grouping** - _Select classification_:  Can that play a role in the JSON query?

<img width="565" height="862" alt="Image" src="https://github.com/user-attachments/assets/72e8c652-27b1-4f08-b281-2094a455fe0c" />

Like @pitkant mentions here [https://github.com/rOpenGov/pxweb/issues/278#issuecomment-3396483317](https://github.com/rOpenGov/pxweb/issues/278#issuecomment-3396483317), the API on the website gives this filter which looks a bit different than other filters:

<img width="684" height="603" alt="Image" src="https://github.com/user-attachments/assets/9bc3ec12-6085-4335-ab67-39c574f2feb5" />

 Could that be related to this issue?

**Comment by MansMeg** [comment]
Yes. Have you asked the Latvian NSI about this? It looks like an error in their API? The code is generic and works in general. So it looks like an error on their side? Maybe even for that specific dataset?

**Comment by MansMeg** [comment]
Other datasets work, right?

**Comment by mapdek** [comment]
Yes, any other dataset seems to work fine!

**Comment by mapdek** [comment]
ok, I can send them an email to verify this filter for this specific dataset

**Comment by mapdek** [comment]
Hi @MansMeg , the reply from Statistics Latvia was this:

```
**We have tested the RUI020m API endpoint using Postman by submitting a JSON query payload. The API response was returned successfully, and the specified filter operates as expected.
 
In this case, the filter is mandatory because the data table contains two sub-tables, each associated with a separate codelist. Therefore, the filter must explicitly indicate which codelist should be used for retrieving the NACE values. This ensures that the correct classification is applied during query execution.
 
Based on our validation results, the JSON filter structure defined by the API for this table is correct and functions as intended.
 
Please let us know if any additional clarification is required.**
```

Do you think it could be related to the fact that there are these sub-tables for this specific dataset?

**Comment by MansMeg** [comment]
Hmm. Still strange. Does it work for you if you use JSON and do the call?

**Comment by mapdek** [comment]
> Hmm. Still strange. Does it work for you if you use JSON and do the call?

Would you have a sample code how to do it for the JSON call? Or is it what @pitkant indicated:

```
json_query_path <- "path/to/query.json"
pxq2 <- pxweb_query(json_query_path)
px_data <- pxweb_get(url = "https://data.stat.gov.lv/api/v1/en/OSP_PUB/START/NOZ/RU/RUI/RUI020m", query = pxq2)
```

**Comment by MansMeg** [comment]
Exactly, you can also print json query with pxweb interactive if you want.

**Comment by MansMeg** [comment]
I still get this error. The sample has an obsolete NACE_MIG value, but even updated valid requests currently fail for the RUI020* table family, so this looks like a Statistics Latvia table/API-side problem rather than something fixable in pxweb. I close it for now. But feel free to open it up if Stats Latvia claim it is on our end (it usually is not).
